File: StructuralPattern-Zen/src/org/uncle/lee/bridge/origin/RefinedAbstraction.java
package org.uncle.lee.bridge.origin;

import org.uncle.lee.utils.LogUtils;

public class RefinedAbstraction extends Abstraction {
	private static final String TAG = RefinedAbstraction.class.getSimpleName();

	public RefinedAbstraction(Implementor implementor) {
		super(implementor);
	}

	@Override
	public void operate() {
		doBefore();
		super.operate();
		doAfter();
	}

	private void doBefore() {
		LogUtils.d(TAG, "doBefore");
	}

	private void doAfter() {
		LogUtils.d(TAG, "doAfter");
	}

}

File: Section5/src/structural/bridge/RefinedAbstraction.kt
package structural.bridge

class RefinedAbstraction(implementor: Implementor) : Abstraction(implementor) {
    override fun operation() {
        doJobA()
        super.operation()
        doJobB()
    }

    private fun doJobA() {
        println("do Job A")
    }

    private fun doJobB() {
        println("do Job b")
    }
}